src/lib/input/input-container.scss

```diff
@@ -3,70 +3,124 @@
+    ($mat-input-subscript-margin-top + $mat-input-line-height) * $mat-input-helper-font-scale;
+// The amount of y-offset to apply to the placeholder when it is floating.
+$mat-input-floating-placeholder-offset: -$mat-input-infix-padding - 1em;
 
-// Gradient for showing the dashed line when the input is disabled.
-$mat-input-underline-disabled-background-image:
-    linear-gradient(to right, rgba(0, 0, 0, 0.26) 0%, rgba(0, 0, 0, 0.26) 33%, transparent 0%);
 
 // Applies a floating placeholder above the input itself.
 @mixin mat-input-placeholder-floating {
   display: block;
-  transform: translate3d(0, -1.35em, 0) scale($mat-input-floating-placeholder-scale-factor);
-  width: 100% / $mat-input-floating-placeholder-scale-factor;
+  transform: translateY(-$mat-input-infix-margin-top - $mat-input-infix-padding)
+             perspective(1px) // fixes text blurriness.
+             scale($mat-input-helper-font-scale);
+  width: 100% / $mat-input-helper-font-scale;
 }
 
+
 .mat-input-container {
   display: inline-block;
   position: relative;
   font-family: $mat-font-family;
-  line-height: normal;
+
+  // TODO(mmalerba): should this come from typography settings? If so a lot of this CSS needs to go
+  // into some sort of typography mixin.
+  line-height: $mat-input-line-height;
 
   // To avoid problems with text-align.
   text-align: left;
 
   [dir='rtl'] & {
     text-align: right;
   }
-
-  // Allow icons in a prefix/suffix/hint/etc to adapt to the correct size.
-  & .mat-icon,
-  & .mat-datepicker-toggle {
-    width: 1em;
-    height: 1em;
-    font-size: 100%;
-    vertical-align: top;
-  }
 }
 
 // Global wrapper. We need to apply margin to the element for spacing, but
 // cannot apply it to the host element directly.
 .mat-input-wrapper {
-  margin: $mat-input-wrapper-spacing 0;
-  // Account for the underline which has 4px of margin + 2px of border.
-  padding-bottom: 6px;
+  position: relative;
+  padding-bottom: $mat-input-wrapper-padding-bottom;
 }
 
-// We use a table layout to baseline align the prefix and suffix classes.
-// The underline is outside of it so it can cover all of the elements under
-// this table.
-// Flex does not respect the baseline. What we really want is akin to a table
-// as want an inline-block where elements don't wrap.
-.mat-input-table {
-  display: inline-table;
-  flex-flow: column;
-  vertical-align: bottom;
+// We use a flex layout to baseline align the prefix and suffix elements.
+// The underline is outside of it so it can cover all of the elements under this flex container.
+.mat-input-flex {
+  display: inline-flex;
+  align-items: baseline;
   width: 100%;
+}
+
+.mat-input-prefix,
+.mat-input-suffix {
+  white-space: nowrap;
+  flex: none;
 
-  & > * {
-    display: table-cell;
+  // Allow icons in a prefix or suffix to adapt to the correct size.
+  .mat-icon,
+  .mat-datepicker-toggle {
+    width: 1em;
+    height: 1em;
+    font-size: $mat-input-prefix-suffix-icon-font-size;
```

Comment:
Actually this one I want to depend on the input's font-size so that the ratio from the spec is preserved. Therefore doesn't need to come from the typography API